lib/oban/instance.ex

```diff
@@ -0,0 +1,224 @@
+defmodule Oban.Instance do
+  @moduledoc """
+  Provides a simplified interface to the `Oban` library.
+  """
+
+  @doc """
+  It creates a facade for the `Oban` functions.
+
+  It allows you to avoid having to pass the `t:Oban.name/0` value to the `Oban` functions, as it is automatically set to
+  the module name. As well as allowing you to configure the `Oban` instance in the application's configuration using
+  the module name under a given OTP application key.
+
+  ## Examples
+
+  ### In an application module
+
+      defmodule MyApp.Oban do
+        use Oban.Instance,
+          otp_app: :my_app,
+          repo: MyApp.Repo
+      end
+
+  Now you can register the `MyApp.Oban` module in the application's supervision tree:
+
+      defmodule MyApp.Application do
+        use Application
+
+        def start(_type, _args) do
+          children = [
+            MyApp.Repo,
+            MyApp.Oban
+          ]
+
+          opts = [strategy: :one_for_one, name: MyApp.Supervisor]
+          Supervisor.start_link(children, opts)
+        end
+      end
+
+  ### Avoiding the need of passing the `Oban` instance
+
+  Instead of calling `Oban` functions passing the `Oban` instance, you can use the `Oban.Instance` module directly:
+
+      Oban.Instance.insert(MyApp.MyJob.new(args: %{ "field" => "value" }))
+  """
+  defmacro __using__(opts \\ []) do
+    {opts, child_opts} = Keyword.split(opts, [:otp_app])
+    otp_app = Keyword.fetch!(opts, :otp_app)
+
+    quote do
+      @doc """
+      Returns the Oban child spec for the application. This should be added to the application's supervision tree.
+
+      The `:name` option is ignored and set to `#{inspect(__MODULE__)}`.
+
+      ## Examples
+
+          defmodule MyApp.Application do
+            use Application
+
+            children = [
+              {#{inspect(__MODULE__)}, prefix: "special"}
+            ]
+
+            opts = [strategy: :one_for_one, name: MyApp.Supervisor]
+            Supervisor.start_link(children, opts)
+          end
+      """
+      def child_spec(opts) do
+        unquote(child_opts)
+        |> Keyword.merge(Application.get_env(unquote(otp_app), __MODULE__, []))
```

Comment:
I'm unsure how to resolve the combination of application config, child spec options, and module-level options, but I'd like something similar to `Ecto.Repo.init/1` that allows people to break complex config out of `runtime.exs` or `config.exs`.

For example, with a callback like this:

```elixir
  def init(opts) do
    base = [
      engine: Oban.Pro.Queue.SmartEngine,
      repo: MyApp.Repo,
      plugins: plugins()
    ]

    Keyword.merge(base, opts)
  end

  defp plugins do
    [
      {Oban.Plugins.Cron, crontab: crontab()},
      {Oban.Plugins.Gossip, []},
      {Oban.Pro.Plugins.DynamicLifeline, []},
      {Oban.Pro.Plugins.DynamicPruner, pruner_overrides()},
      {Oban.Pro.Plugins.DynamicQueues, queues: queues()},
      {Oban.Pro.Plugins.DynamicScaler, scalers: scalers()},
      {Oban.Web.Plugins.Stats, []}
    ]
  end
```

Reply:
My two cents, I would wait until people develop use cases that grant such an `init` callback.

Ecto was done before `runtime.exs`, so it makes sense that it is part of it, but nowadays; (at least personally), I haven't found much need for such a style since people could use `runtime.exs` to read from the environment variables if that is what they want (which for the most part is what the Ecto example shows).

I am not opposed to it, but I rather wait, the worst part is coming up with a naming and how it suppose to work but we already have a reference.